### ISSUE TITLE
Runtime check

### ISSUE THREAD

**Issue by int-index** [isovector/type-sets#4]
It would be useful to have a matching runtime check:

```
data SOrdering :: Ordering -> Type where
  SGT :: SOrdering 'GT
  SLT :: SOrdering 'LT
  SEQ :: SOrdering 'EQ

cmpType :: TypeRep a -> TypeRep b -> SOrdering (CmpType a)
```

Of course, `TypeRep` already has an `Ord` instance, so we can do runtime comparisons:

```
Prelude Type.Reflection> compare (SomeTypeRep (typeRep @Int)) (SomeTypeRep (typeRep @Bool))
LT
```

The issue is that this result is not guaranteed to match whatever `CmpType` of `type-sets` would produce, as it's the result of comparing the `TypeRep` fingerprints.

**Comment by isovector** [comment]
Can you do this via

```
class LowerOrd (ord :: 'Ordering) where
  lowerOrd :: Ordering

instance LowerOrd 'EQ where
  lowerOrd = EQ

-- etc

cmpType :: forall ord a b. (ord ~ CmpType a b, LowerOrd ord) => TypeRep a -> TypeRep b -> Ordering
cmpType _ _ = lowerOrd @ord
```

?

**Comment by int-index** [comment]
I don't think that's an option. Let's say I store a dynamic map on disk, each element is `(TypeRep a, a)`:

```
(Int, 15)
(Int, 20)
(Bool, True)
(String, "Hello")
```

Then I read from disk and maybe I want to build a type-indexed `Map` from these elements, using their types as keys.

For each element, all I have is its `TypeRep` and the element itself. I don't have `LowerOrd` evidence on disk. And if I was to store it, the amount of evidence would be `O(n^2)` because I'd have to store the comparisons between every pair of elements.

The only way forward here is to make `CmpType` operate on information that is also available from the `TypeRep`.

**Comment by isovector** [comment]
Good point. My current ordering is just lexigraphic, so I'd expect `on compare show` to behave the same on typereps as it does with `CmpType`. I could formalize that!

**Comment by int-index** [comment]
`show` for `TypeRep` does not include the package name or the module name:

```
module A where
data T
```

```
module B where
data T
```

```
ghci> typeRep @A.T
T
ghci> typeRep @B.T
T
ghci> show (typeRep @A.T) == show (typeRep @B.T)
True
```

**Comment by isovector** [comment]
It looks like you should be able to reconstruct this via [`typeRepTyCon`](https://www.stackage.org/haddock/lts-14.0/base-4.12.0.0/Type-Reflection.html#v:typeRepTyCon), and then subsequently `tyConPackage` et al. It'd be nasty to keep in sync, but maybe we could test via quickchecking some GADT that uses both?

**Comment by int-index** [comment]
`TypeRep` are typically compared by their fingerprints. It is a fast comparison. I was wondering if it's possible to compute these fingerprints at compile time and use them in `CmpType` as well?
